File: app/src/main/java/com/lazylee/lzywanandroid/view/divider/ArticleRecycleDivider.java
package com.lazylee.lzywanandroid.view.divider;

import android.graphics.Canvas;
import android.graphics.Rect;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;

import android.support.annotation.ColorInt;
import android.support.v7.widget.RecyclerView;
import android.view.View;

/**
 * divider for home fragment's recycler view
 * Created by lazylee on 2018/4/12.
 */

public class ArticleRecycleDivider extends RecyclerView.ItemDecoration {

    private Drawable mDrawable;
    private int mDividerHeight;

    /**
     * @param color divider color
     */
    public ArticleRecycleDivider(@ColorInt int color) {
        this(color, 2);
    }

    public ArticleRecycleDivider(@ColorInt int color, int height) {
        this.mDrawable = new ColorDrawable(color);
        this.mDividerHeight = height;
    }

    @Override
    public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state) {
        int position = parent.getChildLayoutPosition(view);
        int childCount = parent.getAdapter().getItemCount();

        if (isFirstRow(position)){
            outRect.set(0,mDividerHeight,0,mDividerHeight/2);
        }else if (isLastRow(position,childCount)){
            outRect.set(0,mDividerHeight/2,0,mDividerHeight);
        }else {
            outRect.set(0,mDividerHeight/2,0,mDividerHeight/2);
        }
    }

    @Override
    public void onDraw(Canvas c, RecyclerView parent, RecyclerView.State state) {
        c.save();
        int childCount = parent.getChildCount();
        for (int i = 0; i < childCount; i++) {
            final View child = parent.getChildAt(i);
            final int left = child.getLeft() + child.getPaddingLeft();
            final int top = child.getBottom();
            final int right = child.getRight() - child.getPaddingRight();
            final int bottom = top + mDividerHeight;
            mDrawable.setBounds(left, top, right, bottom);
            mDrawable.draw(c);
        }
        c.restore();
    }

    /**
     * 判断该项是否为第一行
     *
     * @return 是否是第一行
     */
    private boolean isFirstRow(int position) {
        return position == 0;
    }

    /**
     * 判断该项是否是最后一行
     *
     * @return 是否是最后一行
     */
    private boolean isLastRow(int position, int childCount) {
        return position == childCount - 1;
    }

    private int getDividerHeight() {
        return mDividerHeight;
    }
}

File: app/src/main/java/com/lazylee/lzywanandroid/ui/view/divider/ArticleRecycleDivider.kt
package com.lazylee.lzywanandroid.ui.view.divider

import android.graphics.Canvas
import android.graphics.Rect
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable

import androidx.annotation.ColorInt
import androidx.recyclerview.widget.RecyclerView

import android.view.View

/**
 * divider for home fragment's recycler view
 * Created by lazylee on 2018/4/12.
 */

class ArticleRecycleDivider @JvmOverloads constructor(@ColorInt color: Int, private val dividerHeight: Int = 2) : RecyclerView.ItemDecoration() {

    private val mDrawable: Drawable

    init {
        this.mDrawable = ColorDrawable(color)
    }

    override fun getItemOffsets(outRect: Rect, view: View, parent: RecyclerView,
                                state: RecyclerView.State) {
        val position = parent.getChildLayoutPosition(view)
        val childCount = parent.adapter!!.itemCount

        if (isFirstRow(position)) {
            outRect.set(0, dividerHeight, 0, dividerHeight / 2)
        } else if (isLastRow(position, childCount)) {
            outRect.set(0, dividerHeight / 2, 0, dividerHeight)
        } else {
            outRect.set(0, dividerHeight / 2, 0, dividerHeight / 2)
        }
    }

    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        c.save()
        val childCount = parent.childCount
        for (i in 0 until childCount) {
            val child = parent.getChildAt(i)
            val left = child.left + child.paddingLeft
            val top = child.bottom
            val right = child.right - child.paddingRight
            val bottom = top + dividerHeight
            mDrawable.setBounds(left, top, right, bottom)
            mDrawable.draw(c)
        }
        c.restore()
    }

    /**
     * 判断该项是否为第一行
     *
     * @return 是否是第一行
     */
    private fun isFirstRow(position: Int): Boolean {
        return position == 0
    }

    /**
     * 判断该项是否是最后一行
     *
     * @return 是否是最后一行
     */
    private fun isLastRow(position: Int, childCount: Int): Boolean {
        return position == childCount - 1
    }
}
/**
 * @param color divider color
 */
